### PR TITLE
gh-105979: Fix exception handling in `unmarshal_frozen_code` (`Python/import.c`)

### LINKED ISSUE #105979: Assertion failure in `_imp.get_frozen_object` if data object contains bad marshal data
Repro:
```python
import _imp
>>> _imp.get_frozen_object('x', b"6\'\xd5Cu\x12")
```
Output:
```
python: Objects/object.c:541: PyObject_Repr: Assertion `!_PyErr_Occurred(tstate)' failed.
Aborted (core dumped)
```
I'll submit a PR.

### Linked PRs
* gh-105980
* gh-106055
* gh-106100

### PR DESCRIPTION
**Pull request by chgnrdv**
Fixes #105979

Clear exception that is set in `PyMarshal_ReadObjectFromString` when it fails to unmarshal given data. It will prevent calling `PyUnicode_FromFormat` in `set_frozen_error` with exception set

* Issue: gh-105979

### CODE REVIEW

**Review by kumaraditya303 — APPROVED** [review]
LGTM, please add a short news entry.

File: Misc/NEWS.d/next/Core and Builtins/2023-06-22-19-16-24.gh-issue-105979.TDP2CU.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+Fix crash in :func:`_imp.get_frozen_object` due to improper exception handling.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Core and Builtins/2023-06-22-19-16-24.gh-issue-105979.TDP2CU.rst]:
```suggestion
Fix crash in :func:`!_imp.get_frozen_object` due to improper exception handling.
```

**Comment by chgnrdv** [comment]
@kumaraditya303, sorry for being late, but I can reproduce the issue on 3.11.4 too. Should these changes be backported then?

File: Lib/test/test_import/__init__.py [hunk 2]
```diff
@@ -763,6 +764,13 @@ def test_dll_dependency_import(self):
                                     env=env,
                                     cwd=os.path.dirname(pyexe))
 
+    def test_issue105979(self):
+        # this used to crash
+        with self.assertRaises(ImportError) as cm:
+            _imp.get_frozen_object("x", b"6\'\xd5Cu\x12")
+        self.assertIn("Frozen object named 'x' is invalid",
```
Comment (jhapalak) [inline Lib/test/test_import/__init__.py:771]: Shouldn't assert on the exception message, since they aren't part of the API.

**Pull request merged**